lightning/src/ln/channelmanager.rs

```diff
@@ -1334,6 +1340,47 @@ impl<ChanSigner: ChannelKeys, M: Deref, T: Deref, K: Deref, F: Deref> ChannelMan
 		})
 	}
 
+	/// Generates a signed node_announcement from the given arguments and creates a
+	/// BroadcastNodeAnnouncement event. Note that such messages will be ignored unless peers have
+	/// seen a channel_announcement from us (ie unless we have public channels open).
+	///
+	/// RGB is a node "color" and alias is a printable human-readable string to describe this node
+	/// to humans. They carry no in-protocol meaning.
+	///
+	/// addresses represent the set (possibly empty) of socket addresses on which this node accepts
+	/// incoming connections. These will be broadcast to the network, publicly tying these
+	/// addresses together. If you wish to preserve user privacy, addresses should likely contain
+	/// only Tor Onion addresses.
+	///
+	/// Panics if addresses is absurdly large (more than 500).
+	pub fn broadcast_node_announcement(&self, rgb: [u8; 3], alias: [u8; 32], addresses: Vec<msgs::NetAddress>) {
+		let _ = self.total_consistency_lock.read().unwrap();
+
+		const HALF_MESSAGE_IS_ADDRS: u32 = 64*1024 / (msgs::NetAddress::max_len() as u32 + 1) / 2;
+		debug_assert!(HALF_MESSAGE_IS_ADDRS >= 500, "500 addresses uses up more than half the message!");
+		if addresses.len() > 500 {
+			panic!("More than half the message size was taken up by public addresses!");
+		}
+
+		let announcement = msgs::UnsignedNodeAnnouncement {
+			features: NodeFeatures::supported(),
+			timestamp: self.last_node_announcement_serial.fetch_add(1, Ordering::AcqRel) as u32,
+			node_id: self.get_our_node_id(),
+			rgb, alias, addresses,
+			excess_address_data: Vec::new(),
+			excess_data: Vec::new(),
+		};
+		let msghash = hash_to_message!(&Sha256dHash::hash(&announcement.encode()[..])[..]);
```

Comment:
just to be sure -- spec says the signature should be over the _double_ hash, and this seems to be a single hash?

Reply:
Right, bitcoin_hashes types are confusing. That tiny little d that easy to miss means double :).

Reply:
Ah, it is easy to miss.. :sweat_smile: